### ISSUE TITLE
Search is too slow

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#5176]
Currently the search addon limits itself to 1000 highlights by default. Contrast this to monaco's 20000 results before it gives up highlighting. When we increase our limit to 20000 the search itself ends up blocking the renderer pretty significantly.

Here is a profile using my i7-12700KF 3.61 GHz:

![Image](https://github.com/user-attachments/assets/ba976bed-23d7-456a-8457-fbe209b11dfa)

![Image](https://github.com/user-attachments/assets/4d3fa85c-e4d0-414c-a46c-6d865bbce14d)

Note that the search is also not incremental, so all successive searches take the same amount of time even for non-regex searches.

VS Code issue: https://github.com/microsoft/vscode/issues/230365

**Comment by jerch** [comment]
Hmm I am pretty that monaco makes use of an index - text data in editors tend to stay pretty stable normally beside the few lines the user is currently editing. So even more costly index creation is justified by a rather low change rate over time.

The terminal buffer is not quite the same in that regard - it tends to change text data substantially over time, so the question is, if we can find a low cost index repr, which still can speed up the search significantly.

I have lit. no clue about search addon, so idk how it currently approaches this issue. I prolly would check, if a trie / suffix array can help here, but those are typically quite expensive to create, so might be too much for fast changing data. If thats the case, maybe a lighter version only caching line & content position for the first n letters, then switching to full line content eval is faster. How deep such an index cache should go, depends on the trie branching and the size of the position buckets, which is again data dependent (so there is no overall good solution for that).

**Comment by Tyriar** [comment]
I don't think monaco has a particularly clever index as you're suggesting actually, just caching similar to us (I might be wrong). I think the performance problems mainly stem from the fact that neither of us really got too involved in the search addon and performance wasn't a big focus. Performance of registering so many decorations might be a part of this too, would need to investigate.

**Comment by jerch** [comment]
> I don't think monaco has a particularly clever index as you're suggesting actually...

Ah ok, well that was just a wild guess from my side. But if they do it likewise we the search addon currently, the perf difference might indeed result from another bottleneck and not the actual positional search.

**Comment by jerch** [comment]
Did a few orientation perf tests with your test file to see, where we start from (scrollback set to 100k, searching the full file line count with one go with sync code just collecting buffer positions):
- turn whole buffer into one big string and search with `indexOf('FIXES')`: 72960 matches in **~100 ms**
- search directly on the arraybuffer with codepoints:
  - uncached: 72960 matches in **~55 ms**
  - 1st codepoint cached:
    - 1st run: 72960 matches in **~70 ms**
    - later runs: 72960 matches in **~40 ms**

Seems the upper limit for naive search on the whole buffer is 50-100 ms with Javascript on my machine for your test file. While WASM typically shows a 2-5x speedup for those "chew on that big chunk of data" tasks, I dont expect it to be faster here due to the needed data copies, thus did not test it.

Configuring the current code to also match all occurences takes currently **~6.3 s**:
![image](https://github.com/user-attachments/assets/8e1a5279-dd62-4076-a5c3-2bec3c023b18)
mainly caused by setTimeout/clearTimeout calls:
![image](https://github.com/user-attachments/assets/e7377c88-83b9-4ccc-a3da-7d79e3e29b42)
from these lines: https://github.com/xtermjs/xterm.js/blob/a5fc111d623dccf009508489726959afb87e1011/addons/addon-search/src/SearchAddon.ts#L436-L437. Disabling them it runs at **~800 ms** (~8x faster):
![image](https://github.com/user-attachments/assets/146bfdb4-c06c-4cb2-961f-4d73348d6d11)

Remember that those numbers cannot be directly compared (the current addon code does a lot more than just collecting buffer positions). It still gives some hints about the possible gains. Also I did not test any more clever caching, as the file is a flat repeat of the search string, so any branch cutting on lines wont apply at all.

**Comment by jerch** [comment]
@Tyriar Did a bit more investigation - when I add the highlightAll part to my fast direct buffer search, things dont look good anymore, runtime for 72960 highlighted matches is now **~600 ms**:
- buffer search: 46 ms
- `_createResultDecoration` is at **~470 ms**, mostly resulting from:
  - `registerMarker` **~170 ms**
  - `registerDecoration` **~170 ms**
 
Tracing things further down, big portions of the overhead runtime come from:
- src/vs/base/common/event.ts:L1097 with **~290 ms** (mostly from markers)
- minor GC with **~190 ms** (GC invocations mostly point into vs event codebase)

To me it seems, that the vs event system is too heavy for fast marker creation. Also the results suggest, that the event impl is not specifically optimized for low GC profile:
![image](https://github.com/user-attachments/assets/badf5e3e-e410-429a-a5c5-810c0565cc5e)

registerMarker path
![image](https://github.com/user-attachments/assets/1de119a2-cea7-475c-9953-9cf8a7c75bcd)

registerDecoration path
![image](https://github.com/user-attachments/assets/7eeb8fde-c127-4c54-a560-4660e15a4e86)

I'd suggest to re-eval the usage of vs/eventemitter for terminal markers. Idk if those could be made faster per se, the code looks quite convoluted to me and I guess thats for a reason. Maybe we dont need all the event fanciness for markers and can get away with a less complete but much faster impl instead.

**Comment by jerch** [comment]
More profiling, this time to get behind the high GC:
![image](https://github.com/user-attachments/assets/ec937ea7-fd87-4f5d-b62c-e9139d541b4d)

I had to lower scrollback to 10000 to get this profile loading, it matched 10018x in the buffer and highlighted those. The allocation table looks crazy, 360000 function allocations alone for 10k highlights. Note that all numbers are almost perfect multiples of 10018, so we see here the "workload" per decoration done:
- Function: 36x, ~1700 bytes
- Object: 12x, ~1690 bytes
- Context: 33x, ~1632 bytes
- (JSArrayBufferData: those are the bufferlines, so not newly allocated)
- Array: 4x, ~1465 bytes
- _Marker2: 1x, ~1148 bytes
- UniqueContainer: 8x, ~384 bytes
- Emitter: 3x, ~368 bytes
- Decoration: 1x, ~332 bytes

Now those numbers cannot simply be added to see the full deal for one decoration, as "Context" also accounts arrow "Function"s, and "Object" also contains the bootstrapping setup (thus all bufferlines). What can be said so - the decoration creation is highly dominated by function creations, most of them being arrow functions. When looking into the details under "Function" and "Context" it becomes clear, that mostly the marker attached to a decoration creates that function pressure. Looking at the code all these anon arrow functions have to go through the event emitter setup, which explains, why event.ts:1097 is so high in timeline profile and triggers GC that often.

Long story short - our marker setup placing tons of arrow functions on `onDispose` is very heavy. We prolly should make that much more light-weighted, if we want it to work with 10k decorations.

**Comment by jerch** [comment]
And here the allocation counts for 10k matches:
![image](https://github.com/user-attachments/assets/f5fc6d67-c060-4f69-a079-b3336de7d231)

indicating that, `_event` gets called very often from `addMarker` and `_createResultDecoration`. The high runtime is in the end a result a high call count of that getter creating function objects over and over (also causing high GC pressure).

When looking at event.ts:1097 I wonder, if reshaping the getter with anonymous function creation into a real function would lower the runtime stress a lot here.

**Comment by jerch** [comment]
Last profiling against v5.5.0 (before the vs addition) - finishs in ~400 ms (vs 600 ms in master) and has a less toxic function creation overhead:
![image](https://github.com/user-attachments/assets/8a942d61-bfc2-45c4-a8ef-a8db723618d4)

![image](https://github.com/user-attachments/assets/0780eb57-eb1f-4309-b433-112fa88a9787)

So with the VS event emitter we lost ~50% performance on marker/decoration creation (+200 ms). The original runtime of 400 ms with the old event emitter is still high, and mainly caused by the big stack of single actions at each marker/decoration creation. To fix the latter, the whole marker/decoration impl would be on trial - it either needs batchers to save some runtime in multiple creations at once or even very reduced marker/decoration impls with containers doing the event housekeeping.

**Comment by AnouarTouati** [comment]
I am interested in helping with this.
I have few question/ideas and things I tried already.

1. Why does the cache have TTL ? there are already event listeners to clear the cache onLineFeed onResize and onCursorMove(i dont see why a cursor move should clear the cache ). 
Anyway I removed the TTL part to clear up my profiler's data.
2. I experimented with making the search return a small set. release the main thread to render the result and then go and get the full result. i did this with a setTimeout. It helped a lot with the UX.
3.But when you try to search for another we fall into the same problem of having to clear all markers and it blocks the render again. Can we just set the markers array to an empty array or could there be other markers not related to the search that we must not remove ?
Maybe another solution is to count the matches but only apply the decorators to the visible lines and a few hidden ones on top and bottom ?

**Comment by jerch** [comment]
@AnouarTouati Yes, there can be other markers, thus we cannot simply forget them all at once. The problem kinda boils down to them being too heavy to process in high numbers.

A possible fix might be to batch their release/addition starting at the current terminal view for early feedback, and then do the rest of the buffer by later chunks.

**Comment by AnouarTouati** [comment]
Ok I will try to do that.
I noticed that "incremental" search only serves to do the following :
Suppose you have A B D C D in the terminal 
Search for C to select it
Then search for D and it selects the **second** D. 
Is this intentional ?

**Comment by Tyriar** [comment]
Thanks for looking into this! Commented on the PR.

**Comment by dolonet** [comment]
Came here via [websh](https://github.com/dolonet/websh) — a small browser-based SSH terminal — where we noticed the same freeze on large-scrollback searches. Thanks @jerch for the extremely detailed profile; it makes the bottleneck really clear.

One thing that jumped out re-reading the thread with fresh eyes: the search path and the decoration path are two very different costs stacked on top of each other, and right now they're blended in a single synchronous loop (`_highlightAllMatches` → `createHighlightDecorations` → `registerMarker`/`registerDecoration` × N).

If you wanted to split the problem for a later PR, I'd suggest the order:

1. **Decouple viewport-visible highlights from overview-ruler markers.** Tyriar's concern about losing the ruler density (#5176 comment 2573032523) is real, but the ruler only needs *positions*, not full marker+decoration objects with per-entry event wiring. A thin "ruler position list" pushed straight to the overview ruler would sidestep most of the allocation pressure jerch profiled (the 36×/marker function allocations come from the `onDispose` emitter setup, which the ruler doesn't need).

2. **Then chunk the decoration creation** for the in-viewport set using `requestAnimationFrame`, starting from the viewport and expanding outward. This is the direction @AnouarTouati was going in #5272; the flicker there looked like a sync issue between the old clear and the new add, which becomes much easier to reason about once step 1 has removed most of the marker churn from the hot path.

3. **Incremental narrowing (#5177)** is a purely additive win on top of that: when the new term is a strict extension of the previous one, and options are unchanged and regex is off, re-filtering the cached result list is O(results) instead of O(buffer). The current `incremental` flag doesn't do this — it only shifts the "starting cursor" for next/prev, per @AnouarTouati's finding (#5272 comment 2573754549).

Happy to take a shot at (1) if there's agreement that's the right first slice — it's the lowest-risk piece and unblocks the rest. Didn't want to open a drive-by PR given @AnouarTouati was already engaged on this.

**Comment by PerBothner** [comment]
What stands out as a likely performance problem: Highlighting finds a list of match results, which is turned into a list of decorations. Later, the renderer (at least the dom renderer) calls `forEachDecorationAtCell` for every cell (seemingly twice). And `forEachDecorationAtCell` does a linear scan through the full list of decorations, looking for decorations that overlap the cell.  This is particularly wasteful given that the decorations are stored using a `SortedList`.

Suggestion: The renderer should do a binary search to find the decoration for the first row. Then it searches from that point for decorations matching the current row (and stops for decorations on later rows). On the next row, it updates the starting index in the decorations list.

Am I reading this right?

**Comment by Tyriar** [comment]
@PerBothner looks like that's one of the major problems:

<img width="1335" height="710" alt="Image" src="https://github.com/user-attachments/assets/1a4e186c-0471-48cc-a32e-4d274f58557e" />

Repro'd that with:

1. Launch demo
2. Run ls 4 times in pwsh, ~3 full screen viewports
3. Test > Decorations > Stress Test
4. Scroll

You're right a binary search to find the range of decorations in the viewport would go a long way to solving this problem. More could be done around caching as well since this is called per cell.

**Comment by dolonet** [comment]
Re-read the code on master to double-check — confirms @PerBothner's reading. `DecorationService.forEachDecorationAtCell` walks the full `SortedList` on every call, and it's hit twice per cell by DomRendererRowFactory and also twice per cell (bottom + top layer) by the WebGL `CellColorResolver`, so the cost shows up on both renderers, not just DOM.

Since the list is already keyed on `marker.line`, a fairly small change should cover both @PerBothner's suggestion and the caching @Tyriar mentioned: binary-search once to the first decoration whose range covers the top row, then walk forward and stop when `ymin > row`. And because the renderers iterate cells left-to-right within a row, keeping a row-scoped start index lets all cells on the same row reuse that seek — per-cell work drops to O(1) amortized without a separate cache layer.

One wrinkle worth flagging: tall decorations (`height > 1`) start before the row they cover, so the binary-search target isn't the row itself but row minus max-height-seen — easy to track on insert/remove.

Happy to take a stab at this as a small PR if it's useful. It's orthogonal to the search-addon-side work we discussed earlier, and would help any app with lots of decorations. No worries if you'd rather keep it in-house or shape the API first — just let us know.

**Comment by PerBothner** [comment]
A complication is that a decorator can cover multiple lines. It's difficult for a sorted list to efficiently handle the case "give me the set of decorations that overlap a given cell".

An idea: Create one entry in the decorations array for each line that the decoration covers. I.e assume we have a decoration D start starts at line L and ends at line L+N-1. In that case we create N-1 helper objects:
```
class ContinuedDecoration extends IInternalDecoration {
   realDecocation: IInternalDecoration;
   offset: number; // from 1 to N-1
}
```
The sort key in the SortedList becomes:
```
e => e instanceof ContinuedDecoration ? e.realDecoration.marker.line + e.offset : e.marker.line
```
Iterating over decorations that overlap a row is now: Use binary search to find the first decoration that overlaps the row. If a ContinuedDecoration, deal with that. Continue forwards in the array until the sort key is greater than the end row (or only row).

Dealing with ContinuedDecoration needs to avoid duplicates. We can deal with that by keeping a (presumably short) list of realDecocation instances we've already handled.

We probably need some hooks so DecorationService can get lower-level access to the SortedList. Making a copy with `values` is needlessly expensive.

**Comment by PerBothner** [comment]
Note @dolonet:
> so the binary-search target isn't the row itself but row minus max-height-seen — easy to track on insert/remove.

I don't think that works in the general case where decorations can overlap in random ways.

**Comment by PerBothner** [comment]
A more radical change:

In the following I assume PR #5797's idea of a LogicalLine, which contains the cell contents of a line assuming infinite screen width; a visible BufferLine becomes a sub-range of a LogicalLine. This idea doesn't require PR #5797 - but it works better with it.

Each LogicalLine points to a list (array or linked list) of Markers associated with the LogicalLine.

The Marker does not have a line number. Therefore, it does not need to be updated on a window resize.

A Marker may have a range of cells it applies to. This is specified as a start and end column number, assuming an infinitely-wide terminal. (Therefore these column ranges do not need to be updated on terminal resize.)

There is no global list of Markers associated with a terminal. To iterate over all Markers, you iterate over all lines, then each Marker associated with the line.

The Marker class can be extended. For example addons may create their own kind of Marker, and then attach the Marker object to a LogicalLine. Possible sub-classes include url links, shell integration regions, search highlighting, the selection, and images.

We might consider getting rid of ExtendedAttrs and just using markers.

A Decoration is kind of Marker that affects rendering. For simplicity, I assume a Decoration covers only a single line. One way to handle decorations that span multiple lines (such as for images), is to use multiple decorations that share state.  Or we can can allow multiple LogicalLine objects to point to the same decoration.

There is no global list of decorations. Neither rendering or serialization need it - they just need to find the markers associated with the LogicalLines they are working on. When a LogicalLine is disposed (perhaps because it is scrolled away), we dispose of the associated markers.

Among other benefits, we solve the "slow search problem" - or at least a major part of it. Checking for decorations while rendering becomes much cheaper. For lines with no markers, there is nothing to search. Lines will seldom have more than one or two markers, so searching that list is trivial.

I may have forgotten some issues. For example, there might be places where a Marker not having a line number is complication, though right now I can't think of any.

This should probably be a separate issue, if people think there is merit in the idea. It would probably break compatibility, and is definitely not a small local change.

**Comment by dolonet** [comment]
Thanks, that's fair — max-height-seen is correct but not tight, and one tall decoration collapses the binary search back near the start of the list. ContinuedDecoration is the right shape for a general fix.

Two small observations on top.

On dedup: within a single forEachDecorationAtCell(x, row) call each covering decoration appears exactly once in the sorted list for that row — either as the real entry at marker.line === row, or as a ContinuedDecoration with offset === row − L. So the "already handled" list is only really needed if a viewport-range scan API is added later; the current per-cell callsites don't need it. Might be worth keeping in mind when shaping the SortedList hook so we don't pay for dedup we don't need.

On SortedList.values(): it does `return [...this._array].values()` every call, i.e. a full N-element array copy. forEachDecorationAtCell is invoked twice per cell in DomRendererRowFactory (the isDecorated check and the color-override loop) and twice per cell in CellColorResolver (bottom and top layer). At 10k decorations and an 80×24 viewport that's roughly 154M element copies per frame before any actual scan work. Dropping to `this._array` directly (or exposing a range iterator) looks like a cheap win regardless of which algorithmic approach ends up landing for the range query.

If the maintainers want to split this, one reasonable slicing might be:

1. Add a range-iterator hook on SortedList (e.g. valuesInRange(keyMin, keyMax), or an index-based seek) that avoids the array copy.
2. Use it in forEachDecorationAtCell / getDecorationsAtCell with a fast path for single-line decorations and a linear fallback if any live decoration has height > 1.
3. ContinuedDecoration as a follow-up when a clean answer is wanted for multi-line cases (ruler markers, shell-integration regions, etc.).

That way the cheap, low-risk perf win (the values() copy and the per-cell scan for the single-line / search-addon case) can land independently of the larger data-model question, which still wants maintainer input on the SortedList API shape.

@Tyriar, happy to prototype the first step as a small PR if that slicing sounds reasonable, or hold off if you'd rather shape the ContinuedDecoration-ready API in one go. Didn't want to front-run the design.

Re: logical-line-owned markers — agree that's a separate issue, not going to go into it here.

**Comment by PerBothner** [comment]
While [this proposal](https://github.com/xtermjs/xterm.js/issues/5176#issuecomment-4308682755) may be overkill if you just want to speed up search, note that making Markers be extensible is probably something we want to do anyway, as discussed in issue #5807. And if we are going to use Markers more extensively, having them be owned by the LineBuffer/LogicalLine makes a lot of sense.

So my recommendation is speed up search if you can so do with a fairly simple change, but perhaps not put a huge amount of effort into it.

**Comment by dolonet** [comment]
Agreed on scoping down. Will park ContinuedDecoration and the search-addon-side chunking until the #5807 / Marker-on-LogicalLine direction settles — both touch the same data model PerBothner is reshaping, and neither is needed for the per-cell hot path that's actually pegging the renderer today.

Plan for a small PR:

1. Add a range accessor on `SortedList` that avoids the `[...this._array].values()` copy — either `valuesInRange(keyMin, keyMax)` returning an iterator, or exposing a `seek(key) → index` so callers can walk `_array` directly. Both are easy; the second is a bit more flexible for the row-scoped reuse below.
2. In `DecorationService.forEachDecorationAtCell`: binary-search to the first decoration whose row range covers the requested row, walk forward, stop when `marker.line > row`. Single-line fast path; if any live decoration has `height > 1`, fall back to the current linear scan (tracked with a counter on add/remove so the fast path is the default).
3. Row-scoped start index reuse so the four per-cell callsites (DomRendererRowFactory ×2, CellColorResolver ×2) on a given row share one seek — per-cell work becomes O(1) amortized for the common case.

No API surface change beyond the new `SortedList` method, no new decoration types, nothing that constrains the larger Marker rework.

@Tyriar — before I start, do you have a preference between `valuesInRange(min, max)` and an index-based `seek(key)` on `SortedList`? Iterator is cleaner at the callsite; index is what step 3 actually wants. Happy to go either way, or expose both if you'd rather keep the iterator as the public shape and `seek` internal.

**Comment by PerBothner** [comment]
In PR #5853 I created a proof-of-concept implementation of [above idea](https://github.com/xtermjs/xterm.js/issues/5176#issuecomment-4308682755). At this point I think it would be more productive to follow up on #5853 (which builds on PR #5797) than work on more localized tweaks. But I'm biased ...

**Comment by dolonet** [comment]
Nice, took a look at #5853 — agree the LogicalLine-owned-markers shape is the right long-term answer, and `forEachDecorationAtCellLine` cleanly avoids the per-cell scan rather than just speeding it up. Worth noting the dependency chain though: #5853 builds on #5797, which is itself a 17-file / ~850-line refactor of the buffer model still under review, and #5853 is WIP with lifetime/dispose, test regressions, the WebGL renderer, and multi-line payloads still open. Realistically that's at least one release cycle out, probably two.

The small `SortedList` range-accessor + binary-search change is ~30 lines, touches only `DecorationService` and `SortedList`, and is purely additive — it doesn't constrain the data model and gets thrown away cleanly when `forEachDecorationAtCell` itself goes away under #5853. The win for users running on master in the meantime is the same per-cell hot path that #5853 fixes.

@Tyriar — happy to either prototype the small PR now as a stopgap, or stand down and wait for the #5797 → #5853 stack to land. Your call; I don't want to add review load if you'd rather concentrate it on the bigger refactor.

**Comment by PerBothner** [comment]
[This suggestion](https://github.com/xtermjs/xterm.js/issues/5176#issuecomment-4308682755) is part of PR #5853. It would be helpful if someone did some measurements/testing of how well that PR solves the problem.

**Comment by dolonet** [comment]
Took a stab at measuring #5853 against master, using `test.txt` from the issue (72,960 `FIXES` lines), 100k scrollback, headless Chromium + Playwright, DOM renderer only. Medians of 5 runs.

| `highlightLimit` | `search` highlight-all (ms) | scroll frame p50 (ms) | scroll frame p99 (ms) |
|---|---|---|---|
| **1,000** (default) | master 22 / #5853 **46** (2.1×) | master 16.7 / #5853 16.6 | master 19.5 / #5853 22.2 |
| **5,000** | master 67 / #5853 **373** (5.6×) | master 16.8 / #5853 **156** (9.3×) | master 28.7 / #5853 **166** (5.8×) |
| **20,000** (Monaco's cap) | master 202 / #5853 **4,748** (24×) | master 38.5 / #5853 **2,055** (53×) | master 69.5 / #5853 **2,173** (31×) |
| **100,000** (all 72,960 matches) | master 706 / #5853 **120,224** (170×) | master 800 / #5853 — | master 940 / #5853 — |

At 100k the render path on #5853 hangs the headless renderer (Chromium's hang monitor kills the page mid-scroll); the search-side number is creation-only, n=5. All other cells are n=5 full runs.

#5853 is faster than master in exactly one regime — `highlightLimit ≤ 1000` where neither has work to do — and gets worse than master from 5k onward in both highlight creation *and* scroll-frame render.

The hot spot looks like `marker.line` going from a public `O(1)` field on master to an `O(buffer.lines.length)` linear scan in the new getter (`src/common/buffer/Marker.ts`). Every caller of `decoration.marker.line` is now doing a full buffer walk per access, and there are several on hot paths:

- `DecorationService` — `_decorations = new SortedList(e => e?.marker.line, ...)`, so each `insert()` calls the getter `O(log N)` times via the binary-search comparator. N inserts of 20k decorations into a 100k-line buffer ≈ 20,000 × log₂(20,000) × 100,000 ≈ 2.8e10 ops, which matches the 4.7 s creation we see.
- `BufferDecorationRenderer` (lines 78, 92) — `element.style.top = (decoration.marker.line - ydisp) * cellHeight`, called for every live decoration on every scroll, so the scroll-frame cost is `decorations × buffer_lines` per frame. 5,000 × 100,000 ≈ 5e8 per frame ≈ 150 ms, again matching.
- `ColorZoneStore` (lines 66–89) — same getter, four times per decoration on every ruler refresh.
- `addons/addon-search/src/DecorationManager.ts` (lines 87, 140) — `_highlightedLines.add(decoration.marker.line)` and `_highlightedLines.has(marker.line)` once per match.

So the concept of `forEachDecorationAtCellLine` is fine — at `highlightLimit=1000` scroll frames stay on 60 fps just like master — but the data-model swap that comes with it makes every `marker.line` lookup `O(buffer)`, and that's hit on enough hot paths that the PR ends up slower than master at any count where #5176 actually bites.

What would unblock measurement is making `marker.line` cheap again: either cache `(line, col)` on `Marker` and invalidate on the events that already move markers, or maintain a back-link `LogicalLine → first buffer-line index` so the getter is one indirection. With that, the four call sites above stop being the bottleneck and we can see whether `forEachDecorationAtCellLine` itself delivers the per-cell win the PR is built around.

While that lands — and especially given #5853 is stacked on #5797 — the small `SortedList` range-accessor + binary-search change I sketched [in the slicing above](https://github.com/xtermjs/xterm.js/issues/5176#issuecomment-4321678481) (~30 lines, touches only `DecorationService` and `SortedList`, purely additive, gets thrown away cleanly when #5853 lands) would still help users on master in the meantime. Happy to send that as a PR if it's useful; just say the word.

Bench harness and raw JSON: https://gist.github.com/dolonet/312727975dd3e117c15edb8ee0732cc3.

**Comment by dolonet** [comment]
Following up on my own comment — I should not have speculated on the renderer-drop cause (it's our outer timeout, not Chromium's hang monitor) and the `marker.line` getter actually lives in the #5797 commits this PR is stacked on, not in #5853 itself. Apologies for the loose phrasing.

More importantly, I ran a CDP CPU profile through Playwright to actually attribute the cost instead of arguing it from code-reading. `--js-flags=--expose-gc --disable-hang-monitor`, 100 µs sample interval, 5000 highlights into the same `test.txt`, DOM renderer.

**Create phase** (highlight-all, 695 ms total):
| % self | self (ms) | function |
|---|---|---|
| 78.0 | 542 | `Marker.get line` (multiple deopt entries) |
| 2.7 | 19 | `CircularList.get` |
| 1.7 | 11 | `CircularList._getCyclicIndex` |
| 0.5 | 4 | `DecorationManager._createResultDecorations` |
| 0.3 | 2 | `BufferLine.logical` |

**Scroll phase** (20 frames, 1854 ms total):
| % self | self (ms) | function |
|---|---|---|
| 77.1 | 1429 | `Marker.get line` |
| 8.8 | 163 | `CircularList.get` |
| 7.8 | 145 | `CircularList._getCyclicIndex` |
| 0.0 | 0.2 | `DecorationService.forEachDecorationAtCellLine` |

So the concept of `forEachDecorationAtCellLine` is validated by the profile — it's essentially free per cell, exactly what the PR set out to achieve. The whole regression is the `marker.line` getter and its inner `buffer.lines.get()` chain, on both the insertion path (SortedList comparator) and the per-decoration-per-frame render path (`BufferDecorationRenderer`). If `marker.line` becomes O(1) again — by any of the routes mentioned earlier or simply by not using `marker.line` on hot paths and threading the `LogicalLine` through instead, the way `forEachDecorationAtCellLine` already does — the PR should deliver the per-cell win it's designed for.

Profile dumps (`.cpuprofile`, DevTools-loadable) added to the same gist: https://gist.github.com/dolonet/312727975dd3e117c15edb8ee0732cc3.

**Comment by PerBothner** [comment]
Oops. Fixing the calls to marker.line in the SortedList is trivial - just don't don't use a SortedList. If using a plain array, then the only non-deprecated use of the `_decorations` list that is not amortized  O(1) appears to be the call to `delete` in the `onDispose` callback created in `registerDecoration`. If we can arrange that `dispose` is predominantly called in reverse order of creation then `delete` does not have to do any data moving. I haven't looked into how difficult that would be - certainly the `reset` method can easily call the `dispose` in optimal order.

Possibly better is make the `_decorations` list be a doubly-linked list.  Then the delete would be O(1) as well. The forward and back links can be in the `Decoration` object itself, reducing the need for an extra object per link, as well as the associated lndirection. The only non-trivial part of this implementing an `IterableIterator` to be returned by the `decorations` method, but that doesn't look difficult.

**Comment by PerBothner** [comment]
I'm thinking it might make sense to store a `_voffset` field in each `BufferLine`. This would be a line number, except it wouldn't get updated when scrolled-out lines are trimmed. (Thus to get an actual line number you would subtract the starting `_voffset`.) This should be cheap to maintain, and would allow efficient calculation of marker.line.

A future refinement would be to store the `_voffset` in pixel units, rather than line numbers: This would enable mixing lines with different heights (perhaps for bigger header lines, images, or vertical separators), and would alo be helpful for smooth scrolling.

**Comment by dolonet** [comment]
`_voffset` lines up with what the CPU profile [I posted](https://github.com/xtermjs/xterm.js/issues/5176#issuecomment-4495531349) points at: `Marker.get line` was ~78% self in the create phase and ~77% in scroll, almost entirely the `buffer.lines.get()` walk. A `_voffset` field on `BufferLine` makes `marker.line` O(1), which is exactly what those two numbers need — so this looks like the right fix.

Two notes on top:

- It covers more than the `SortedList` swap [suggested above](https://github.com/xtermjs/xterm.js/issues/5176#issuecomment-4499568987). Moving `_decorations` to a plain array / linked list removes the *insertion*-side `marker.line` calls (the comparator), but the larger number in the scroll profile is `BufferDecorationRenderer` reading `decoration.marker.line` for every live decoration every frame — a different container doesn't touch that. `_voffset` fixes both at the source, and with an O(1) `marker.line` the `SortedList` is back to its master-branch cost, so it doesn't need replacing for #5176's sake.
- `_voffset` is also cleaner than the per-`Marker` `(line, col)` cache I [floated earlier](https://github.com/xtermjs/xterm.js/issues/5176#issuecomment-4494739224): it needs no invalidation on `lineFeed` or scrollback trim — a new line just takes the next value, and trimmed lines are defined not to update it — so there's no event wiring to get wrong.

One thing worth confirming: reflow. On a width change `BufferLine`s get re-ranged, so the `_voffset` sequence has to be rebuilt — `O(buffer)` once per resize, which is fine since resize is already `O(buffer)`, but it shouldn't slip through. I haven't traced #5797's reflow path, so I can't say how much falls out for free there.

For #5176 the line-unit version is all that's needed; the pixel-unit refinement is a nice direction for variable line heights / smooth scrolling but is a separable follow-up that shouldn't gate the perf fix.

Happy to re-run the [comparison table](https://github.com/xtermjs/xterm.js/issues/5176#issuecomment-4494739224) (master vs branch, same `test.txt`, harness in the gist) once `_voffset` is in a testable branch — that should isolate whether `forEachDecorationAtCellLine` delivers the per-cell win with `marker.line` off the hot path. Since it touches `BufferLine` itself, though, that's a core-data-model call for @Tyriar / @jerch.

**Comment by PerBothner** [comment]
Hm. Managing a `_voffset` field is conceptually not difficult, but it would require some non-trivial API changes. For example, when `CircularList.set` is called we would need to update the line's `_voffset` - in a number of places. And it can't be done in `CircularList.ts` since that code know nothing of `BufferList` (even though it's not used for anything else). I think it would be a cleaner API to replace `buffer.lines.set(index, line)` by `buffer.setLine(index, line)` (which could also take care of updating the `_voffset` property), but that is a more extensive API change than we'd like at this point.

So I'm leaning towards restoring the explicit `line` property in `Marker`. I think we can do that without the losing the fast `forEachDecorationAtCellLine`.

**Comment by PerBothner** [comment]
I checked in a form of my `_voffset` idea. I believe this makes `Marker.line` O(1) again,

I have not yet removed the use of `SortedList` (the other change we discussed). I'm hoping to get to it tomorrow.

**Comment by PerBothner** [comment]
I also removed the use of `SortedList`, which needlessly sorted decorations based on `Marker.line`.

Testing/benchmarking would be welcome.

**Comment by Tyriar** [comment]
I added decoration indexing by line in https://github.com/xtermjs/xterm.js/pull/5902, this is a more scoped down change that fixes the issue without major refactors. I've also been doing a parallel exploration into improving the find addon algorithm which might get a PR soon.
